### PR TITLE
CircleCI workflow

### PR DESCRIPTION
**Pull request by mikz**
New CircleCI workflow using native docker feature, with parallel jobs, test parallelization, artifacts, ... all the jazz. And finishing within 10 minutes. There will be more PRs, but this is the bare minimum to get all building. Huge kudos to @gsaslis, who worked on this too. I just took all the differences and re-committed it to clean up the history and make it smaller. More changes are coming.

### Fix transactional tests

When executing both transactional and non-transactional tests we got random failures because the database was polluted from the previous run. This was caused by several things, the gist is that now all cukes and minitests should be properly isolated and easy to run in one run, without splitting to several processes. We will verify this in the future and simplify the build.

### Unitfy test folders

Some minitests were in the wrong folder. For example `test/controllers` had a mix of integration and functional tests.

### Disabled APIcast integration tests

We have not tacked this yet, so APIcast tests are disabled. But the plan is to have it started all the time like database and just use the API to control it.

### Test files

Because we need to split test files across containers, we first need to list the test files that are going to be executed (rspec, minitest, cucumber). For minitest there is a rake task that lists test files, for cucumber there is a profile and a formatter to print the file names. Cucumber runs in dry run, that means it does not load rails, so few support files had to be changed to not crash.

### CODE REVIEW

File: features/support/paths.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
-module NavigationHelpers
+World(Module.new do
+  break unless defined?(DeveloperPortal)
```
Comment (hallelujah) [inline features/support/paths.rb]: Just for curiosity. Why do we need to check that?

Comment (hallelujah) [reply features/support/paths.rb]: I got my answer:

>  Cucumber runs in dry run, that means it does not load rails, so few support files had to be changed to not crash

File: features/old/finance/invoices/multitenant.feature [hunk 2]
```diff
@@ -1,3 +1,4 @@
+@wip
```
Comment (guicassolato) [inline features/old/finance/invoices/multitenant.feature]: Let's not forget to revert this. Meanwhile, we may want to add "WIP" to the description of the PR.

Comment (gsaslis) [reply features/old/finance/invoices/multitenant.feature]: good point! 

how about we revert this particular change (i think this was only done to see just how fast the build would be if this alone didn't take ~3mins or so...) and create a separate issue to track actually _fixing_ this?

Comment (mikz) [reply features/old/finance/invoices/multitenant.feature]: I'll revert this until we get coverage report. Then we should easily see if not running this test decreases test coverage or not and is covered by some other tests.

Comment (gsaslis) [reply features/old/finance/invoices/multitenant.feature]: @mikz i think we should get this merged first and then i can open a separate PR with just the coverage, right?

or would you prefer those commits here?

Comment (mikz) [reply features/old/finance/invoices/multitenant.feature]: Definitely a separace PR.

File: lib/tasks/ci/licenses.rake [hunk 3]
```diff
@@ -9,15 +9,15 @@ namespace :ci do
...
+      unless system("BUNDLE_GEMFILE='Gemfile' lib/threescale_license_finder/bin/threescale_license_finder")
         STDERR.puts "*** License compliance test failed  ***"
         exit 1
       end
     end
     desc 'Generates a report with the dependencies and their licenses'
     task :report do
       STDOUT.puts 'Generating report...'
-      exec("BUNDLE_GEMFILE='Gemfile.on_prem' lib/threescale_license_finder/bin/threescale_license_finder report --format=xml > #{Rails.root.join('doc/licenses/licenses.xml')}")
+      exec("BUNDLE_GEMFILE='Gemfile' lib/threescale_license_finder/bin/threescale_license_finder report --format=xml > #{Rails.root.join('doc/licenses/licenses.xml')}")
```
Comment (didierofrivia) [inline lib/tasks/ci/licenses.rake]: This is going to generate the report with "saas" gems, do we want that?

Comment (mikz) [reply lib/tasks/ci/licenses.rake]: Honestly, I have no idea. But you can't generate licenses for what you don't install and here we are not installing the other gemfile.
Once https://github.com/3scale/porta/pull/44 is merged the "on prem" will be the main gemfile and it should all work.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+#      - v1-gateway-{{ arch }}-{{ .Branch }}-{{ .Revision }}
+
+
+
+attach-to-workspace: &attach-to-workspace
+  attach_workspace:
+    at: .
+
+system-builder: &system-builder
+  image: quay.io/gsaslis/3scale-system-builder:latest
```
Comment (mayorova) [inline .circleci/config.yml]: Shouldn't this be pushed to `quay.io/3scale` instead? @gsaslis @mikz

Comment (mikz) [reply .circleci/config.yml]: Yes, it will be at some point.

**Pull request merged**